### PR TITLE
Added FutureWarning to Series.str.__iter__

### LINKED ISSUE #28277: `series.str.cat(series.str)` is concatenating only the largest string
#### Code Sample, a copy-pastable example if possible

```python
import pandas as pd

arr = ["AbC", "de", "FGHI", "j", "kLLLm"]

ps = pd.Series(arr)
expect = ps.str.cat(others=ps.str)
print(expect)
Out[16]:
0           NaN
1           NaN
2           NaN
3           NaN
4    kLLLmkLLLm
dtype: object
```

#### Problem description
`series.str.cat(series.str)` is concatenating only the largest string in the series but it should concatenate all the strings element wise.

#### Expected Output
```python
Out[18]:
0        AbCAbC
1          dede
2      FGHIFGHI
3            jj
4    kLLLmkLLLm
dtype: object
```

#### Output of ``pd.show_versions()``

<details>

In [2]: pandas.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.7.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.10.0-862.14.4.el7.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: None
LOCALE: en_US.UTF-8

pandas: 0.24.2
pytest: 5.1.2
pip: 19.2.3
setuptools: 41.2.0
Cython: 0.29.13
numpy: 1.17.1
scipy: None
pyarrow: 0.14.1
xarray: None
IPython: 7.8.0
sphinx: 2.2.0
patsy: None
dateutil: 2.8.0
pytz: 2019.2
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml.etree: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10.1
s3fs: 0.3.4
fastparquet: None
pandas_gbq: None
pandas_datareader: None
gcsfs: None

</details>

### PR DESCRIPTION
**Pull request by SaturnFromTitan**
- [x] #28277
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/strings.py [hunk 1]
```diff
@@ -2093,6 +2093,11 @@ def __getitem__(self, key):
             return self.get(key)
 
     def __iter__(self):
+        warnings.warn(
+            "Columnar iteration over characters will be deprecated in future releases.",
+            DeprecationWarning,
```
Comment (WillAyd) [inline pandas/core/strings.py]: Can you raise a `FutureWarning` instead of a `DeprecationWarning`?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: pandas/core/strings.py (hunk 1 above)
Comment (SaturnFromTitan) [reply pandas/core/strings.py]: ✅

File: doc/source/whatsnew/v1.0.0.rst [hunk 2]
```diff
@@ -260,6 +260,31 @@ The following methods now also correctly output values for unobserved categories
    df.groupby(["cat_1", "cat_2"], observed=False)["value"].count()
 
 
+Avoid iterating over ``Series.str``
```
Comment (WillAyd) [inline doc/source/whatsnew/v1.0.0.rst]: So this is a very niche / questionable application to begin with, so probably don’t need a dedicated section for this. Can you just make this a one-liner in the `Deprecations` section that Series.str.__iter__ has been deprecated and will be removed in a future version?

**Review by WillAyd — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v1.0.0.rst (hunk 2 above)
Comment (SaturnFromTitan) [reply doc/source/whatsnew/v1.0.0.rst]: Sure, no problem

**Review by jreback — CHANGES_REQUESTED** [review]
why exactly do we need to deprecate this? I would simply fix the bug

**Comment by SaturnFromTitan** [comment]
@jreback The deprecation follows from the discussion in #28319 where an actual fix to the issue was suggested:

WillAyd:
> I think the best approach would be to deprecate the __iter__ method of Series.str as noted above and remove in the future

**Comment by jreback** [comment]
> @jreback The deprecation follows from the discussion in #28319 where an actual fix to the issue was suggested:
> 
> WillAyd:
> 
> > I think the best approach would be to deprecate the **iter** method of Series.str as noted above and remove in the future

ok i see the conversation and agree deprecation is ok. but this doesnt' fix the actual bug #28277 
 right? (meaning that should be a separate PR).

**Comment by SaturnFromTitan** [comment]
Yes, the bug isn't fixed by that, but @TomAugspurger suggested the expected behaviour should be to raise an exception:
> series.str is an accessor, not an array-like of strings. I would expect an exception to be raised here.

So even though the bug isn't fixed immediately, in later versions, when this deprecation is taking effet, it will be resolved. Do you mean we should keep the issue open until then @jreback?

**Comment by SaturnFromTitan** [comment]
Resolved merge conflict and removed the "closes" tag from the PR description @jreback

**Comment by jreback** [comment]
@SaturnFromTitan can you merge master

**Review by WillAyd — APPROVED** [review]
lgtm

**Comment by SaturnFromTitan** [comment]
CI is green @jreback

**Comment by jreback** [comment]
great will look soon

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @SaturnFromTitan

**Pull request merged**